src/vsr/checkpoint_trailer.zig

```diff
@@ -167,6 +163,24 @@
             return @intCast(stdx.div_ceil(trailer.size, chunk_size_max));
         }
 
+        pub fn chunks(trailer: *Self, trailer_size: u64) []const []align(@sizeOf(u256)) u8 {
+            const chunk_count: u32 = @intCast(stdx.div_ceil(trailer_size, chunk_size_max));
+            for (
+                trailer.block_bodies[0..chunk_count],
+                trailer.blocks[0..chunk_count],
+                0..,
+            ) |*block_body, block, block_index| {
+                const chunk_size = Chunk.size(.{
+                    .block_index = @intCast(block_index),
+                    .block_count = chunk_count,
+                    .trailer_size = trailer_size,
+                });
```

Comment:
I am wondering whether it makes sense to try to slice this by `chunk_size`, rather than by
statically known `chunk_size_max` --- the resulting chunks are only "true" during decoding. When
encoding, we conservatively size chunks, and the last one might not be full.

No strong feelings either way!

Another thing here is that, ideally, we'd compute `StreamChecksum` as we are filling in the chunks,
rather than as a separate path afterwards, so that we don't do two passes over the same data.
Though, given that we have one chunk anyway, not that it matters :)

Reply:
> I am wondering whether it makes sense to try to slice this by chunk_size, rather than by
> statically known chunk_size_max --- the resulting chunks are only "true" during decoding. When
> encoding, we conservatively size chunks, and the last one might not be full.

I don't understand what you mean by this. :thinking:

Reply:
Yeah, sorry, let me rephrase!

What I mean is that the code _looks_ like it slices the blocks into the chunks we actually are going to use. So, for every block, we return not just `block[@sizeOf(vsr.Header)..block_size_max]`, but cap that to the exact size computed by the `Chunk`.


But that's not actually true during encoding! Chunks are allocated conservatively, so encoder might not actually be able to use up all of the chunks. So, for the encoding step, `block_body.* = block[@sizeOf(vsr.Header)..][0..chunk_size];` is actually dead code --- the encoder disregards `chunk_size` info, and instead stops when it itself thinks it should to stop. 

This seems a bit confusing to me. Specifically, the fact that, after encoding, some memory in `chunks` might actually be uninitialized. 

The alternative here is to say that `chunks` doesn't even attempt to compute which fraction of each block is used, and returns a `[]const [chunk_size_max]u8`, leaving it up to the caller to track the actual usage per chunk.

Reply:
Aha, that makes sense, thank you!

I think I'll split `.chunks()` into two different functions, `encode_chunks` and `decode_chunks`. Then neither need a `size` passed in -- `encode_chunks` returns an array of all max-size chunks, and `decode_chunks` uses `trailer.size` to truncate the last chunk. And `decode_chunks` is only used for reading existing chunks, so the returned slice can be `const`.

Reply:
Ah, brilliant, I was only noticing something fishy here, but indeed it looks like what we have here is just two different function!